lib/plug/conn/status.ex

```diff
@@ -3,72 +3,97 @@ defmodule Plug.Conn.Status do
-    request_uri_too_long: 414,
-    unsupported_media_type: 415,
-    requested_range_not_satisfiable: 416,
-    expectation_failed: 417,
-    im_a_teapot: 418,
-    misdirected_request: 421,
-    unprocessable_entity: 422,
-    locked: 423,
-    failed_dependency: 424,
-    upgrade_required: 426,
-    precondition_required: 428,
-    too_many_requests: 429,
-    request_header_fields_too_large: 431,
-    internal_server_error: 500,
-    not_implemented: 501,
-    bad_gateway: 502,
-    service_unavailable: 503,
-    gateway_timeout: 504,
-    http_version_not_supported: 505,
-    variant_also_negotiates: 506,
-    insufficient_storage: 507,
-    loop_detected: 508,
-    not_extended: 510,
-    network_authentication_required: 511
-  ]
-
-  doc = Enum.map(statuses, fn {atom, code} ->
-    "  * `#{inspect atom}` - #{code}\n"
-  end)
+  custom_statuses = Application.get_env(:plug, :statuses, %{})
+
+  statuses = %{
+    100 => "Continue",
+    101 => "Switching Protocols",
+    102 => "Processing",
+    200 => "OK",
+    201 => "Created",
+    202 => "Accepted",
+    203 => "Non-Authoritative Information",
+    204 => "No Content",
+    205 => "Reset Content",
+    206 => "Partial Content",
+    207 => "Multi-Status",
+    208 => "Already Reported",
+    226 => "IM Used",
+    300 => "Multiple Choices",
+    301 => "Moved Permanently",
+    302 => "Found",
+    303 => "See Other",
+    304 => "Not Modified",
+    305 => "Use Proxy",
+    306 => "Switch Proxy",
+    307 => "Temporary Redirect",
+    308 => "Permanent Redirect",
+    400 => "Bad Request",
+    401 => "Unauthorized",
+    402 => "Payment Required",
+    403 => "Forbidden",
+    404 => "Not Found",
+    405 => "Method Not Allowed",
+    406 => "Not Acceptable",
+    407 => "Proxy Authentication Required",
+    408 => "Request Timeout",
+    409 => "Conflict",
+    410 => "Gone",
+    411 => "Length Required",
+    412 => "Precondition Failed",
+    413 => "Request Entity Too Large",
+    414 => "Request-URI Too Long",
+    415 => "Unsupported Media Type",
+    416 => "Requested Range Not Satisfiable",
+    417 => "Expectation Failed",
+    418 => "I'm a teapot",
+    421 => "Misdirected Request",
+    422 => "Unprocessable Entity",
+    423 => "Locked",
+    424 => "Failed Dependency",
+    425 => "Unordered Collection",
+    426 => "Upgrade Required",
+    428 => "Precondition Required",
+    429 => "Too Many Requests",
+    431 => "Request Header Fields Too Large",
+    500 => "Internal Server Error",
+    501 => "Not Implemented",
+    502 => "Bad Gateway",
+    503 => "Service Unavailable",
+    504 => "Gateway Timeout",
+    505 => "HTTP Version Not Supported",
+    506 => "Variant Also Negotiates",
+    507 => "Insufficient Storage",
+    508 => "Loop Detected",
+    510 => "Not Extended",
+    511 => "Network Authentication Required"
+  }
+
+  reason_phrase_to_atom = fn reason_phrase ->
+    reason_phrase
+    |> String.downcase()
+    |> String.replace(~r/[']/, "")
```

Comment:
@lexmag The cases aren't reliable enough for `Macro.underscore/1` some of them don't uppercase the words ("I'm a teapot") and have special characters in them. The input could also be user defined. 

```
iex(3)> "Proxy Authentication Required" |> Macro.underscore   
"proxy _authentication _required"
iex(4)> "I'm a teapot" |> Macro.underscore
"i'm a teapot
```